hoomd/mpcd/ParticleData.cc

```diff
@@ -449,89 +463,118 @@ void mpcd::ParticleData::takeSnapshot(mpcd::ParticleDataSnapshot& snapshot,
         for (unsigned int idx = 0; idx < m_N; ++idx)
             {
-            pos[idx] = make_scalar3(h_pos.data[idx].x, h_pos.data[idx].y, h_pos.data[idx].z);
-            vel[idx] = make_scalar3(h_vel.data[idx].x, h_vel.data[idx].y, h_vel.data[idx].z);
-            type[idx] = __scalar_as_int(h_pos.data[idx].w);
-            tag[idx] = h_tag.data[idx];
+            mpcd::detail::pdata_element particle;
+            particle.pos = h_pos.data[idx];
+            particle.vel = h_vel.data[idx];
+            particle.tag = h_tag.data[idx];
+            particle.comm_flag = 0;
+            particles[idx] = particle;
             }
 
-        // Create per-processor arrays to gather the data back to root
-        std::vector<std::vector<Scalar3>> pos_proc;       // Position array of every processor
-        std::vector<std::vector<Scalar3>> vel_proc;       // Velocities array of every processor
-        std::vector<std::vector<unsigned int>> type_proc; // Particle types array of every processor
-        std::vector<std::vector<unsigned int>> tag_proc;  // Tag array of every processor
-
-        // resize to number of ranks in communicator
-        const MPI_Comm mpi_comm = m_exec_conf->getMPICommunicator();
-        const unsigned int n_ranks = m_exec_conf->getNRanks();
-        const unsigned int rank = m_exec_conf->getRank();
-        pos_proc.resize(n_ranks);
-        vel_proc.resize(n_ranks);
-        type_proc.resize(n_ranks);
-        tag_proc.resize(n_ranks);
+        // size particles per rank and displacements
+        std::vector<int> num_per_rank(n_ranks);       // number per rank
+        std::vector<int> rank_displacements(n_ranks); // displacement of particles per rank
+        MPI_Gather(&m_N, 1, MPI_UNSIGNED, num_per_rank.data(), 1, MPI_UNSIGNED, root, mpi_comm);
+        rank_displacements[0] = 0;
+        for (size_t i = 1; i < num_per_rank.size(); ++i)
+            {
+            rank_displacements[i] = rank_displacements[i - 1] + num_per_rank[i - 1];
+            }
 
-        // collect all particle data on the root processor
-        const unsigned int root = 0;
-        gather_v(pos, pos_proc, root, mpi_comm);
-        gather_v(vel, vel_proc, root, mpi_comm);
-        gather_v(type, type_proc, root, mpi_comm);
-        gather_v(tag, tag_proc, root, mpi_comm);
+        // gather data back to root rank
+        MPI_Gatherv((rank != root) ? particles.data() : MPI_IN_PLACE,
+                    m_N,
+                    m_mpi_pdata_element,
+                    particles.data(),
+                    num_per_rank.data(),
+                    rank_displacements.data(),
+                    m_mpi_pdata_element,
+                    root,
+                    mpi_comm);
 
         if (rank == root)
             {
+            /*
+             * First sort the particles into tag order. This is probably a slow step, but could be
+             * sped up by sorting on each rank first then merging the sorted segments here.
```

Comment:
I can make this optimization in a subsequent PR, as well as packing the buffers on the GPU. I left it off for now in order to get this finalized.

Reply:
OK, no problem. You might find the `GatherTagOrder` class helpful in implementing the optimization. I wrote it to solve a similar problem when writing GSD files and gathering forces.
https://github.com/glotzerlab/hoomd-blue/blob/f725ae3b796da9d1128c0388fcf36111bcdd781c/hoomd/HOOMDMPI.h#L511-L531

Reply:
Thanks! I remembered that you had written this and wanted to try it, but didn’t have time now. I will keep this mind when I come back to it later. Gathering these snapshots can be very slow.